drv/lpc55-rng/src/main.rs

```diff
@@ -9,119 +9,124 @@
 #![no_std]
 #![no_main]
 
-use core::mem::size_of;
-use drv_lpc55_syscon_api::{Peripheral, Syscon};
+mod config;
+
+use core::{cmp, usize};
+use drv_lpc55_syscon_api::Syscon;
 use drv_rng_api::RngError;
 use idol_runtime::{ClientError, NotificationHandler, RequestError};
+use lib_dice::{persistid_cert_tmpl::SUBJECT_CN_LENGTH, RngSeed, SeedBuf};
+use lib_lpc55_rng::Lpc55Rng;
 use rand_chacha::ChaCha20Rng;
-use rand_core::block::{BlockRng, BlockRngCore};
 use rand_core::{impls, Error, RngCore, SeedableRng};
+use ringbuf::ringbuf;
+use sha3::{
+    digest::crypto_common::{generic_array::GenericArray, OutputSizeUser},
+    digest::FixedOutputReset,
+    Digest, Sha3_256,
+};
 use userlib::task_slot;
+use zeroize::Zeroizing;
 
-use lpc55_pac as device;
-
-task_slot!(SYSCON, syscon_driver);
-
-struct Lpc55Core {
-    pmc: &'static lpc55_pac::pmc::RegisterBlock,
-    rng: &'static lpc55_pac::rng::RegisterBlock,
-    syscon: Syscon,
-}
+cfg_if::cfg_if! {
+    if #[cfg(feature = "dice-seed")] {
+        use config::DataRegion;
+        use hubpack::SerializedSize;
+        use lib_dice::{persistid_cert_tmpl::SUBJECT_CN_RANGE, CertData, RngData};
+        use ringbuf::ringbuf_entry;
+        use serde::Deserialize;
+        use stage0_handoff::{HandoffData, HandoffDataLoadError};
+        use userlib::UnwrapLite;
 
-impl Lpc55Core {
-    fn new() -> Self {
-        let syscon = SYSCON.get_task_id();
-        Lpc55Core {
-            pmc: unsafe { &*device::PMC::ptr() },
-            rng: unsafe { &*device::RNG::ptr() },
-            syscon: Syscon::from(syscon),
+        // This file is generated by the crate build.rs. It contains instances
+        // of config::DataRegion structs describing regions of memory
+        // configured & exposed to this task by the hubris build.
+        mod build {
+            include!(concat!(env!("OUT_DIR"), "/rng-config.rs"));
         }
-    }
-}
-
-impl BlockRngCore for Lpc55Core {
-    type Item = u32;
-    type Results = [u32; 1];
 
-    fn generate(&mut self, results: &mut Self::Results) {
-        results[0] = self.rng.random_number.read().bits();
+        use build::{DICE_CERTS, DICE_RNG};
     }
 }
 
-struct Lpc55Rng(BlockRng<Lpc55Core>);
-
-impl Lpc55Rng {
-    fn new() -> Self {
-        Lpc55Rng(BlockRng::new(Lpc55Core::new()))
-    }
-
-    fn init(&self) {
-        self.0
-            .core
-            .pmc
-            .pdruncfg0
-            .modify(|_, w| w.pden_rng().poweredon());
-
-        self.0.core.syscon.enable_clock(Peripheral::Rng);
+task_slot!(SYSCON, syscon_driver);
 
-        self.0.core.syscon.enter_reset(Peripheral::Rng);
-        self.0.core.syscon.leave_reset(Peripheral::Rng);
-    }
+#[derive(Copy, Clone, PartialEq)]
+enum Trace {
+    #[cfg(feature = "dice-seed")]
+    NoDiceSeed,
+    #[cfg(feature = "dice-seed")]
+    HandoffError(HandoffDataLoadError),
+    #[cfg(feature = "dice-seed")]
+    NoSeedPersonalization,
+    None,
```

Comment:
AFAIK the position in the enum of the `None` variant shouldn't matter. But no reason it can't be at position 0 and I think it makes better sense as well so I've made this change.

`ringbuf` doesn't care that the enum changes and so long as `humility` is able to get symbols provided w/ the archive it should be fine. Removing the feature cfg stuff from these enum variants would probably only cost a few extra bytes and we can ignore the compilers warnings easy enough if we don't like mutating the type.